[sparse] semi-structured sparse + torch.compile support

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111049

Summary:

This PR adds in torch.compile support for semi-structured sparsity,
using the subclass tracing @bdhirsh added.

Based on wether we are using cuSPARSELt or CUTLASS, we return a
different representation of the inner tensors.

Test Plan:
```
python test/test_sparse_semi_structured.py -k compile
```

Reviewers:

Subscribers:

Tasks:

Tags:

jcaip committed:
[wip] semi-structured sparse + torch.compile support

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[wip] semi-structured sparse + torch.compile support"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
So I think we will need some additional work before subclass tracing and 2:4 sparsity can work together. Loosely I think the problem is the op level at which inductor breaks stuff down to call the subclass is more decomposed than in eager (which we have support for). 

For example - fp16 `addmm` in eager is a simple swap with `torch._cslt_sparse_mm`, but in the compiled version it looks like it's trying to break down `addmm` into  smaller pieces, including casts to fp32 for computation type. See here for a trace: https://www.internalfb.com/phabricator/paste/view/P851580855

It looks like when compiling, inductor is trying different combinations of ops to represent `aten.linear`
```
transpose -> addmm
permute(0,1) -> addmm
reinterpret_tensor(flip strides) -> addmm.out (instead of addmm.default)
```

When I try a `linear -> contiguous -> relu` I also see it trying to run `mm` and `mm.out`.

So in order to support this, we'll either need to 
1) Restrict the decompositions/codegen somehow? Not sure exactly what the issue here so will need to dig deeper. It's not clear to me what part of the codebase is trying different combinations. 
2) Add op support for all of the possible generated ops. This is probably something we should do anyways, as some of these combinations may be faster. For example, adding support for passing in the output matrix to sparse matmul. 

permute(0,1) and reinterpret_tensor should be mapped to transpose. 

fp32 casts can be no-ops.

jcaip committed:
Update on "[wip] semi-structured sparse + torch.compile support"


Summary:

Placeholder PR for subclassing + 2:4 sparsity

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented on torch/sparse/semi_structured.py:
```diff
@@ -304,6 +337,55 @@ def __torch_dispatch__(cls, func, types, args, kwargs) -> Any:
                 transposed=not args[0].transposed,
             )
 
+        if func is torch.ops.prims.convert_element_type.default:
```
@bdhirsh These are the ops I added to get this test to "work", you can comment these out to see where it fails earlier in the chain.

jcaip committed:
Update on "[wip] semi-structured sparse + torch.compile support"


Summary:

Placeholder PR for subclassing + 2:4 sparsity

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[wip] semi-structured sparse + torch.compile support"


Summary:

Placeholder PR for subclassing + 2:4 sparsity

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[wip] semi-structured sparse + torch.compile support"


Summary:

Placeholder PR for subclassing + 2:4 sparsity

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[wip] semi-structured sparse + torch.compile support"


Summary:

Placeholder PR for subclassing + 2:4 sparsity

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[wip] semi-structured sparse + torch.compile support"


Summary:

Placeholder PR for subclassing + 2:4 sparsity

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[wip] semi-structured sparse + torch.compile support"


Summary:

Placeholder PR for subclassing + 2:4 sparsity

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[wip] semi-structured sparse + torch.compile support"


Summary:

Placeholder PR for subclassing + 2:4 sparsity

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] semi-structured sparse + torch.compile support"


Summary:

This PR adds in torch.compile support for semi-structured sparsity,
using the subclass tracing bdhirsh added.

Based on wether we are using cuSPARSELt or CUTLASS, we return a
different representation of the inner tensors.

Test Plan:
```
python test/test_sparse_semi_structured.py -k compile
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

cpuhrsch reviewed (commented):
I'll let @bdhirsh have the final word

jcaip commented:
@pytorchbot merge

jcaip commented:
@pytorchbot merge

DanilBaibak commented:
@pytorchbot revert -m "Broken trunk" -c nosignal

DanilBaibak commented:
Hi @jcaip! Unfortunately, I need to revert your PR because it broke some periodic and slow jobs. Here you can find more details - https://hud.pytorch.org/pytorch/pytorch/commit/ac02531babab028cb260d2225ff9e91e92df063b

bdhirsh commented on torch/sparse/semi_structured.py:
```diff
@@ -213,6 +213,37 @@ def __init__(
+        if self.compressed_tensor_cusparselt is not None:
+            return ['compressed_tensor_cusparselt'], (self.original_shape, self.transposed)
+        else:
+            return ['sparse_tensor_cutlass', 'meta_tensor_cutlass'], (self.original_shape, self.transposed)
```
nit: you might want to defensively save the `compressed_tensor_cusparselt` bool in the meta for extra checking (assert that when it is false there are two inner tensors, otherwise one). But probably doesn't matter too much.

bdhirsh commented on test/test_sparse_semi_structured.py:
```diff
@@ -341,6 +341,44 @@ def test_mlp(self, device, dense_input_shape, backend):
+            (1) The sparse tensor subclass should turn nn.Linear into `aten._structured_sparse_linear` + `aten.contiguous()`
+            (2) Inductor should fuse the .contiguous() call into the relu
+        """
+        SparseSemiStructuredTensor._FORCE_CUTLASS = backend == "cutlass"
```
should we also have a test for cusparselt?

jcaip replied:
I don't think we test cuSPARSELt in CI but this should test both, the test is parameterized by backend (cutlass, cusparselt).

bdhirsh commented on test/test_sparse_semi_structured.py:
```diff
@@ -341,6 +341,44 @@ def test_mlp(self, device, dense_input_shape, backend):
 
+    @parametrize("backend", SEMI_STRUCTURED_SUPPORTED_BACKENDS)
+    @parametrize("dense_input_shape", [(1, 128), (64, 128), (128, 128), (64, 128, 128)])
+    def test_mlp_contiguous_relu_compile(self, dense_input_shape, backend, device):
```
oh for the failing trunk CI - you probably need to mark this test as skipped when running in windows CI:

```
from torch.testing._internal.common_utils import IS_WINDOWS
 @unittest.skipIf(IS_WINDOWS, "torch.compile doesn't work with windows")
```

jcaip committed:
Update on "[sparse] semi-structured sparse + torch.compile support"


Summary:

This PR adds in torch.compile support for semi-structured sparsity,
using the subclass tracing bdhirsh added.

Based on wether we are using cuSPARSELt or CUTLASS, we return a
different representation of the inner tensors.

Test Plan:
```
python test/test_sparse_semi_structured.py -k compile
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
@pytorchbot merge

jcaip commented:
@pytorchbot merge

jcaip commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "Sorry I'm pretty sure this caused a memory leak https://hud.pytorch.org/pytorch/pytorch/commit/408f210938176870133a3dde5e8fbc4926cafbc0 https://github.com/pytorch/pytorch/actions/runs/6550388354/job/17790615103 `test_sparse_semi_structured.py::TestSparseSemiStructuredCUDA::test_mlp_contiguous_relu_compile_backend_cutlass_dense_input_shape_(1, 128)_cuda - RuntimeError: CUDA driver API confirmed a leak in __main__.TestSparseSemiStructuredCUDA.test_mlp_contiguous_relu_compile_backend_cutlass_dense_input_shape_(1, 128)_cuda! Caching allocator allocated memory was 235008 and is now reported as 352256 on device 0. CUDA driver allocated memory was 359333888 and is now 361431040.`" -c nosignal

jcaip commented:
This is kind of strange. 

I was able to reproduce this error running locally, but I don't think it's related to my changes here. 

Running the following command yields failures for the torch.compile test, note that the other cusparselt/cutlass tests do not report a memory leak. 
```
PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_sparse_semi_structured.py

............................................................................................
======================================================================
ERROR: test_conversions_all_patterns_backend_cutlass_cuda_float16 (__main__.TestSparseSemiStructuredCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2452, in wrapper
    with policy():
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 1902, in __exit__
    raise RuntimeError(msg)
RuntimeError: CUDA driver API confirmed a leak in __main__.TestSparseSemiStructuredCUDA.test_conversions_all_patterns_backend_cutlass_cuda_float16! Caching allocator allocated memory was 1536 and is now reported as 2560 on device 0. CUDA driver allocated memory was 1044054016 and is now 1046151168.

To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_sparse_semi_structured.py -k test_conversions_all_patterns_backend_cutlass_cuda_float16

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

======================================================================
ERROR: test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(1, 128)_cuda (__main__.TestSparseSemiStructuredCUDA)
Test nn.Linear + .contiguous() + nn.ReLU with SparseSemiStructuredTensor + torch.compile
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2452, in wrapper
    with policy():
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 1902, in __exit__
    raise RuntimeError(msg)
RuntimeError: CUDA driver API confirmed a leak in __main__.TestSparseSemiStructuredCUDA.test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(1, 128)_cuda! Caching allocator allocated memory was 2048 and is now reported as 53760 on device 0. CUDA driver allocated memory was 1438318592 and is now 1440415744.

To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_sparse_semi_structured.py -k test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(1, 128)_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
Ran 168 tests in 110.646s

FAILED (errors=2)
```
Then I tried commenting out the [call](https://github.com/pytorch/pytorch/pull/111049/files#diff-b7b57bc1e34145ef89c7929751d5d26aeecc8edfb37da9c60e9d3f0a1335133cR377) to `to_sparse_semi_structured.py`, which means we are just compare torch.compile vs eager mode. But I still see the compile tests fail. strangely I see more failures?
```
............................................................................................
======================================================================
ERROR: test_conversions_all_patterns_backend_cutlass_cuda_float16 (__main__.TestSparseSemiStructuredCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2452, in wrapper
    with policy():
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 1902, in __exit__
    raise RuntimeError(msg)
RuntimeError: CUDA driver API confirmed a leak in __main__.TestSparseSemiStructuredCUDA.test_conversions_all_patterns_backend_cutlass_cuda_float16! Caching allocator allocated memory was 1536 and is now reported as 2560 on device 0. CUDA driver allocated memory was 1044054016 and is now 1046151168.

To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_sparse_semi_structured.py -k test_conversions_all_patterns_backend_cutlass_cuda_float16

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

======================================================================
ERROR: test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(1, 128)_cuda (__main__.TestSparseSemiStructuredCUDA)
Test nn.Linear + .contiguous() + nn.ReLU with SparseSemiStructuredTensor + torch.compile
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2452, in wrapper
    with policy():
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 1902, in __exit__
    raise RuntimeError(msg)
RuntimeError: CUDA driver API confirmed a leak in __main__.TestSparseSemiStructuredCUDA.test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(1, 128)_cuda! Caching allocator allocated memory was 2048 and is now reported as 35328 on device 0. CUDA driver allocated memory was 1438318592 and is now 1440415744.

To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_sparse_semi_structured.py -k test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(1, 128)_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

======================================================================
ERROR: test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(128, 128)_cuda (__main__.TestSparseSemiStructuredCUDA)
Test nn.Linear + .contiguous() + nn.ReLU with SparseSemiStructuredTensor + torch.compile
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2452, in wrapper
    with policy():
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 1902, in __exit__
    raise RuntimeError(msg)
RuntimeError: CUDA driver API confirmed a leak in __main__.TestSparseSemiStructuredCUDA.test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(128, 128)_cuda! Caching allocator allocated memory was 35328 and is now reported as 68608 on device 0. CUDA driver allocated memory was 1440415744 and is now 1442512896.

To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_sparse_semi_structured.py -k test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(128, 128)_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

======================================================================
ERROR: test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(64, 128)_cuda (__main__.TestSparseSemiStructuredCUDA)
Test nn.Linear + .contiguous() + nn.ReLU with SparseSemiStructuredTensor + torch.compile
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2452, in wrapper
    with policy():
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 1902, in __exit__
    raise RuntimeError(msg)
RuntimeError: CUDA driver API confirmed a leak in __main__.TestSparseSemiStructuredCUDA.test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(64, 128)_cuda! Caching allocator allocated memory was 68608 and is now reported as 101888 on device 0. CUDA driver allocated memory was 1442512896 and is now 1444610048.
                                                                                                                                                                                                                                                                                          [0/1936]
To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_sparse_semi_structured.py -k test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(64, 128)_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

======================================================================
ERROR: test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(64, 128, 128)_cuda (__main__.TestSparseSemiStructuredCUDA)
Test nn.Linear + .contiguous() + nn.ReLU with SparseSemiStructuredTensor + torch.compile
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2452, in wrapper
    with policy():
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 1902, in __exit__
    raise RuntimeError(msg)
RuntimeError: CUDA driver API confirmed a leak in __main__.TestSparseSemiStructuredCUDA.test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(64, 128, 128)_cuda! Caching allocator allocated memory was 101888 and is now reported as 135168 on device 0. CUDA driver allocated
memory was 1444610048 and is now 1446707200.

To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_sparse_semi_structured.py -k test_mlp_contiguous_relu_compile_backend_cusparselt_dense_input_shape_(64, 128, 128)_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

======================================================================
ERROR: test_mlp_contiguous_relu_compile_backend_cutlass_dense_input_shape_(128, 128)_cuda (__main__.TestSparseSemiStructuredCUDA)
Test nn.Linear + .contiguous() + nn.ReLU with SparseSemiStructuredTensor + torch.compile
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2452, in wrapper
    with policy():
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 1902, in __exit__
    raise RuntimeError(msg)
RuntimeError: CUDA driver API confirmed a leak in __main__.TestSparseSemiStructuredCUDA.test_mlp_contiguous_relu_compile_backend_cutlass_dense_input_shape_(128, 128)_cuda! Caching allocator allocated memory was 167424 and is now reported as 200704 on device 0. CUDA driver allocated memory
was 1446707200 and is now 1448804352.

To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_sparse_semi_structured.py -k test_mlp_contiguous_relu_compile_backend_cutlass_dense_input_shape_(128, 128)_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

======================================================================
ERROR: test_mlp_contiguous_relu_compile_backend_cutlass_dense_input_shape_(64, 128)_cuda (__main__.TestSparseSemiStructuredCUDA)
Test nn.Linear + .contiguous() + nn.ReLU with SparseSemiStructuredTensor + torch.compile
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2453, in wrapper
    method(*args, **kwargs)
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 2452, in wrapper
    with policy():
  File "/home/jessecai/local/A/pytorch/torch/testing/_internal/common_utils.py", line 1902, in __exit__
    raise RuntimeError(msg)
RuntimeError: CUDA driver API confirmed a leak in __main__.TestSparseSemiStructuredCUDA.test_mlp_contiguous_relu_compile_backend_cutlass_dense_input_shape_(64, 128)_cuda! Caching allocator allocated memory was 200704 and is now reported as 233984 on device 0. CUDA driver allocated memory w
as 1448804352 and is now 1450901504.

To execute this test, run the following from the base repo dir:
    PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 PYTORCH_TEST_WITH_SLOW_GRADCHECK=1 python test/test_sparse_semi_structured.py -k test_mlp_contiguous_relu_compile_backend_cutlass_dense_input_shape_(64, 128)_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
Ran 168 tests in 119.071s

FAILED (errors=7)
```
cc @clee2000 @bdhirsh do y'all have any ideas what might be causing this? Could it be something funky with the tooling / torch.compile?

bdhirsh commented:
Hey @jcaip - I messed around with the test locally, and this seems like a memory leak directly in cutlass / the SparseSemiStructuredTensor subclass. Here's a minimal repro:
```
import torch
import random
from torch.sparse.semi_structured import SparseSemiStructuredTensor, to_sparse_semi_structured

SparseSemiStructuredTensor._FORCE_CUTLASS = True
def f():
    mask_entries = [random.choice([[0, 1], [1, 0]]) for i in range(16384)]
    A = torch.tensor(mask_entries, dtype=torch.float16, device='cuda').reshape(128, 256).contiguous()
    A_sparse = to_sparse_semi_structured(A)

print(torch.cuda.memory_allocated())
f()
print(torch.cuda.memory_allocated())
```
prints:
```
0
512
```

cpuhrsch commented:
@alexsamardzic - Alek can you take a look into this as well?

alexsamardzic commented:
This is sort of known problem; namely, the code doing conversion is actually `@torch.compiled`-d PyTorch code, and seemingly the memory leak appears because of this.  Namely, the leak will disappear if in `torch/sparse/_semi_structured_conversions.py`, line:
```
def sparse_semi_structured_from_dense_cutlass(dense, compile=True):
```
changed to:
```
def sparse_semi_structured_from_dense_cutlass(dense, compile=False):
```

While trying to change the function in question to come up with minimum self-contained example, I came up with something even unrelated:
```
import torch

@torch.compile
def foo(A):
    return A

def f():
    A = torch.ones(128, dtype=torch.float16, device='cuda')
    B = foo(A)

print(torch.cuda.memory_allocated())
f()
print(torch.cuda.memory_allocated())
```
 It will also print `512` instead of `0`. 

@lezcano @peterbell10

jcaip commented:
@alexsamardzic What do you think about turning off this flag to false now that we have torch.compile support for subclasses? I think this can also cause issues if we call to torch.compile again later.

alexsamardzic commented:
That's fine with me.

jcaip committed:
Update on "[sparse] semi-structured sparse + torch.compile support"


Summary:

This PR adds in torch.compile support for semi-structured sparsity,
using the subclass tracing bdhirsh added.

Based on wether we are using cuSPARSELt or CUTLASS, we return a
different representation of the inner tensors.

Test Plan:
```
python test/test_sparse_semi_structured.py -k compile
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse] semi-structured sparse + torch.compile support"


Summary:

This PR adds in torch.compile support for semi-structured sparsity,
using the subclass tracing bdhirsh added.

Based on wether we are using cuSPARSELt or CUTLASS, we return a
different representation of the inner tensors.

Test Plan:
```
python test/test_sparse_semi_structured.py -k compile
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
Ah I believe, this is because I was not using `torch._dynamo.test_case.TestCase` for my tests

jcaip committed:
Update on "[sparse] semi-structured sparse + torch.compile support"


Summary:

This PR adds in torch.compile support for semi-structured sparsity,
using the subclass tracing bdhirsh added.

Based on wether we are using cuSPARSELt or CUTLASS, we return a
different representation of the inner tensors.

Test Plan:
```
python test/test_sparse_semi_structured.py -k compile
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
@pytorchbot merge -f "passing ciflow/slow now and unrelated failing test"
